### PR TITLE
triclustering modules included

### PR DESCRIPTION
**Pull request by fnattino**
First implementation of the tri-clustering in the package. So far it includes only numpy + threading. Tried to follow the same structure as for the co-clustering.

### CODE REVIEW

File: geoclustering/triclustering_numpy.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    # get data matrix size and fixed averages(upload 3D)
+    Y2 = Z.reshape(d, m*n)
+    Gavg2 = Y2.mean()
+
+    # Randomly initialize row and column clustering
+    R = _initialize_clusters(m, k)
+    C1 = _initialize_clusters(n*d, z)
+    C = _initialize_clusters(n, l)
+    Z = _initialize_clusters(d, z)
```
Comment (rogerkuou) [inline geoclustering/triclustering_numpy.py:51]: Consider not overwriting the input matrix Z

File: geoclustering/triclustering_numpy.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    # get data matrix size and fixed averages(upload rows)
+    Y = np.concatenate(Z, axis=1)
+    Gavg = Y.mean()
+
+    # get data matrix size and fixed averages(upload columns)
+    Y1 = np.concatenate(Z, axis=0)
+    Gavg1 = Y1.mean()
+
+    # get data matrix size and fixed averages(upload 3D)
```
Comment (rogerkuou) [inline geoclustering/triclustering_numpy.py:43]: band?

**Comment by fnattino** [comment]
@rogerkuou actually, I have just realised that the three averages computed in the following: 

https://github.com/phenology/geoclustering/blob/c539e07ae43da8300481476d1bb310cdea7c1ebb/geoclustering/triclustering_numpy.py#L35-L45

.. are all identical: the matrix elements are only re-arranged in different 2D matrices. I will fix this.

File: geoclustering/triclustering_numpy.py (hunk 1 above)
Comment (fnattino) [reply geoclustering/triclustering_numpy.py:51]: Thanks for catching this! Bad namings on my side!

File: geoclustering/triclustering_numpy.py (hunk 2 above)
Comment (fnattino) [reply geoclustering/triclustering_numpy.py:43]: Fixed! (also see comment below)

**Comment by fnattino** [comment]
@rogerkuou Another thought after our discussion earlier. I think the initial confusion on having 4 cluster matrices initialised is justified the algorithm can be slightly modified to be more consistent and easier to grasp at first sight by changing the following lines:

https://github.com/phenology/geoclustering/blob/0777d731585ee84edf1269ddbfd58fde2065939d/geoclustering/triclustering_numpy.py#L44-L57

to:

```python

    # Randomly initialize row and column clustering
    R = _initialize_clusters(m, nclusters_row)
    C = _initialize_clusters(n, nclusters_col)
    B = _initialize_clusters(d, nclusters_bnd)

    e, old_e = 2 * errobj, 0
    s = 0
    converged = False

    while (not converged) & (s < niters):
        C1 = np.kron(np.ones((n, 1)), B)
```

In essence, the C1 matrix is constructed inside the loop, and not randomly initialised but constructed by repeating B (in the same way that R1 and C2 are obtained by repeating R and C). Also, the use of `np.kron` can be replaced `np.tile`, which does the same thing but it seems to be faster and easier to read.

I will fix these.

**Comment by fnattino** [comment]
@rogerkuou Sorry, second thought about the second point above: It is true that the algorithm looks more consistent if C1 is computed from B, but the randomness of the inizialization of C1 might be useful to explore more initial configurations. I will roll back and leave it as it used to be. I will still replace `np.kron` with `np.tiles` though.

**Comment by rogerkuou** [comment]
Hi Francesco, in general I think this is good! Only one thing, I am testing with the following script it is not converging, even if I give a very large `errobj`:

```python
    k = 10 
    l = 5 
    b = 2
    errobj, niters, nruns, epsilon = 100000000, 1, 1000, 10e-8
    Z = np.random.rand(8,1000,20) 
    Z = Z.astype('float64')
    tc = Triclustering(Z, k, l, b, errobj, niters, nruns, epsilon)
    tc.run_with_threads(nthreads=1)

    tc.row_clusters, tc.col_clusters, tc.bnd_clusters
``` 

I have checked it more thoroughly and found maybe it is caused by a hard set of `e_min` in `triclustering.py`. I marked it out in the script. We can discuss this tomorrow.

File: geoclustering/triclustering.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                                self.Z,
+                                self.nclusters_row,
+                                self.nclusters_col,
+                                self.nclusters_bnd,
+                                self.conv_threshold,
+                                self.max_iterations,
+                                self.epsilon):
+                r for r in range(self.nruns)
+            }
+            row_min, col_min, bnd_min, e_min = None, None, None, 0.
```
Comment (rogerkuou) [inline geoclustering/triclustering.py:61]: `e_min` seem to be constantly 0?

Comment (fnattino) [reply geoclustering/triclustering.py:61]: Great that you have caught this! Here I have implicitly assumed the error would be negative - which is not always the case, see your example. A better initialisation of `e_min` is required, not only for the triclustering, but for the coclustering as well. I'll open a new issue.

Comment (fnattino) [reply geoclustering/triclustering.py:61]: I think the reason why we never encountered this issue is that I have always used integer values in the matrices, as taken from the test examples.

**Pull request merged**